### DISCUSSION TITLE
LDM Notes for September 21st, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#6493] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-09-21.md

## Agenda

- Unchampioned issue triage
    - Track subtype exhaustiveness for classes with only private constructors
    - ReadOnlySpan initialization from static data
    - Embedded Language Indicators for raw string literals
    - Implicit Parameters
- Unsafer Unsafeness
    - Conclusion

**Comment by HaloFour** [comment]
> There are some potentially interesting cases in this with ref assemblies as well, since private constructors are normally dropped.

If a type only had private constructors wouldn't dropping it from a reference assembly break the contract since the type would be left with only a public parameterless constructor?

**Reply by svick** [reply]
I think that only applies to C#, not to IL.

**Reply by HaloFour** [reply]
In that case couldn't the compiler consider the lack of any accessible constructors?  I would think that `public`, `protected` and `internal` constructors would need to be included in the reference assembly, so a lack of constructors would indicate only `private` or `private protected` constructors.

**Reply by alrz** [reply]
I think it's "there is no public ctor" vs "there is a private ctor". Relying on the latter would be definitely incorrect.

**Reply by alrz** [reply]
>In that case couldn't the compiler consider the lack of any accessible constructors? 

Accessible from where? The compiler should literally walk the entire compilation to determine that..

**Reply by HaloFour** [reply]
I would think that `public`, `protected` and `internal` constructors would all be within the metadata of that type within a reference assembly.  I guess that does leave `private protected` as a potential gap, though.

Can the rules for reference assemblies be changed so that in this case private constructors must be emitted and the compiler can key off of that?

**Reply by HaloFour** [reply]
Ah, [this](https://learn.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies) would also be a problem:

> If there are no [InternalsVisibleTo](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute) attributes, internal function members are also removed.

Can the rules regarding how these assemblies are emitted be changed?  Could a private constructor, if only just a stub, be considered observable metadata that should be retained in an assembly where a type has nested subtypes?

**Reply by HaloFour** [reply]
I would think that the decisions made here would impact the ability to design "reference DUs", defined by a closed type hierarchy.

**Reply by 333fred** [reply]
These are excellent questions that need to be answered 😉.

**Reply by sab39** [reply]
I always assumed that the default no-argument public constructor was synthesized by the compiler, and that at the IL/runtime level a class might have no constructors at all, in which case there wouldn't be an automatic default one either. There's no reason to emit any constructor at all for a static class, for example. There's nothing saying that a reference assembly has to be expressible in C# source code, is there?

**Reply by HaloFour** [reply]
@sab39 

Right, the problem is that if a reference assembly elides `private protected` and `internal` constructors that it becomes impossible to tell based on that metadata whether those nested types represent a closed hierarchy for the purposes of exhaustiveness checking.
